magda-io/magda
# Split the registry api into "read-only" and "full" for scalability

### What this PR does

Fixes #1649 

This makes the registry change depending on a role - it can either be "full" or "readonly". "full" is the full functionality we know and love, and "readonly" disables webhook processing and all non-get requests.

This means that we can run a single "full" registry-api in prod that handles all changes and the resulting webhooks, while having completely stateless "readonly" registry api pods that can be scaled up and down.

To make this all work I've changed kubernetes so that there's a `registry-api` service like always and also a `registry-api-read-only` service. If the read only registry api is disabled, these both go to the single `registry-api-full` pod. However, if the readonly pods are enabled, then the `registry-api` service will go to the `registry-api-full` pod and the `registry-api-read-only` service will go to the `registry-api-read-only` pod. This means that all communication from inside magda will go to the full registry api pod as usual.

The `registry-api-read-only` service, on the other hand, is used by the gateway - now traffic to `/api/v0/registry/` will go to the `registry-api-read-only` service and hence to `read-only` pods if enabled, otherwise `full` pods. This means that `POST`s etc can no longer be made through the `/v0/api/registry` url which would be annoying, so I've made `/api/v0/registry-auth` always go to a `registry-api-full` pod and still do auth, so that if you sign in in dev you can still copy the cookie and do POSTs/DELETEs/PUTs/PATCHes as usual.

### To Test
Go look at https://github.com/magda-io/magda-config/pull/3/files for a guide to the new config.

Essentially you can either run it as defaults in which case you should have one `registry-api-full` pod which everything uses, internally and externally.

Or you can turn `registry-api.deployments.readOnly.enabled` to `true` in which case you should end up with both a `registry-full` and `registry-read-only` pod which both get different traffic as described above.

### Checklist

-   [x] I've assigned this PR to someone (if you don't know who to assign it to, pick @AlexGilleran)
